musescore: 4.0.2 -> 4.1.0

doronbehar opened the pull request:
Diff: https://github.com/musescore/MuseScore/compare/v4.0.2...v4.1.0

###### Description of changes

Close #243109 .

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

doronbehar committed:
musescore: 4.0.2 -> 4.1.0

Diff: https://github.com/musescore/MuseScore/compare/v4.0.2...v4.1.0

doronbehar commented on pkgs/applications/audio/musescore/default.nix:
```diff
@@ -8,27 +8,19 @@
   cmakeFlags = [
-    "-DMUSESCORE_BUILD_CONFIG=release"
-    # Disable the _usage_ of the `/bin/crashpad_handler` utility. See:
-    # https://github.com/musescore/MuseScore/pull/15577
```
The cmake flags that are mentioned there were renamed, but I agree that some link to more details are useful, I linked the musescore issue instead.

SuperSandro2000 commented on pkgs/applications/audio/musescore/darwin.nix:
```diff
@@ -1,9 +1,9 @@
 
 let
-  versionComponents = [ "4" "0" "1" ];
+  versionComponents = [ "4" "1" "0" ];
```
Why is this not derived from version?

drupol replied:
Curious to know as well.

doronbehar replied:
> Why is this not derived from version?

Good question! I wasn't the one who wrote these expressions in the 1st place. Maybe the darwin expression is not needed anymore... I'll try to test this via ofborg a bit.

doronbehar committed:
musescore: format with nixpkgs-fmt

doronbehar committed:
musescore: Switch to wrapQtAppsHook and stdenv.mkDerivation

The usage of libsForQt5's mkDerivation is deprecated.

doronbehar commented:
@ofborg build musescore

doronbehar commented:
@ofborg build musescore

doronbehar committed:
musescore: Remove unneeded qt deps from buildInputs

doronbehar commented:
@ofborg build musescore

doronbehar commented:
@ofborg build musescore

doronbehar commented:
cc @yurrriq per https://github.com/NixOS/nixpkgs/pull/23204 .

doronbehar commented:
Aish! Syntax errors in /nix/store...-apple-framework-Foundation-11.0.0/Library/Frameworks/Foundation.framework/Headers` ?? How can this be?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/need-help-from-darwin-users-syntax-errors-in-library-frameworks-foundation-framework-headers/30467/1

doronbehar commented:
I'll wait a few days for help from Darwin users, and if none will come, I'll do merge the previous dmg based expression by @yurrriq into `default.nix`.

reckenrode commented:
I’m replying here as well to document (from https://discourse.nixos.org/t/need-help-from-darwin-users-syntax-errors-in-library-frameworks-foundation-framework-headers/30467/2). The issue is portaudio is propagating frameworks from the 10.12 SDK. Mixing frameworks from different SDK versions doesn’t work. It results in errors because the headers aren’t compatible. You need to override portaudio to use the 11.0 SDK.

```nix
let
  # portaudio propagates Darwin frameworks. Rebuild it using the 11.0 stdenv from Qt and the 11.0 SDK frameworks.
  portaudio' = if !stdenv.isDarwin
    then portaudio
    else portaudio.override {
      inherit stdenv;
      inherit (darwin.apple_sdk_11_0.frameworks) AudioUnit AudioToolbox CoreAudio CoreServices Carbon;
    };
in
```

I also want to note this derivation doesn’t build with sandbox enabled. This is a Qt issue. qtbase should be propagating its sandbox profile, but it doesn’t actually seem to work. This is the current workaround I’ve used in a couple of other packages:

```
# HACK `propagatedSandboxProfile` does not appear to actually propagate the sandbox profile from `qtbase`
sandboxProfile = toString qtbase.__propagatedSandboxProfile or null;
```

doronbehar committed:
musescore: Fix darwin build from source

Co-authored-by: Randy Eckenrode <randy@largeandhighquality.com>

doronbehar commented:
@ofborg build musescore

Thanks for the detailed answer @reckenrode ! Hopefully this will work now. I wonder why the default darwin frameworks are not 11_0 throughout all Nixpkgs... Perhaps, `portaudio` and `portmidi` should not propagate these deps?

doronbehar commented:
Still very similar syntax errors...

reckenrode commented:
> I wonder why the default darwin frameworks are not 11_0 throughout all Nixpkgs

The x86_64-darwin SDK hasn’t been bumped in a while. @toonn is working on a bump to 10.13. As for why, I think there is a desire to use a source-based SDK when possible as well as one to support as many users as possible.

I don’t see needing to support multiple SDKs changing. Some software requires a newer one or should be built with a newer one (e.g., Blender) then back deployed (e.g., MoltenVK). Some could be built with a newer one and work okay, but presumably a lot of software does static feature detection instead of dynamic, so we have to build with as old as possible to maximize support. There are definitely some ergonomic and composability issues with the current approach (SDK-specific `callPackages`). I opened #242666 to discuss the future of SDKs on Darwin.

doronbehar commented:
> I opened #242666 to discuss the future of SDKs on Darwin.

Thanks for the link! Do you have any idea what can be done with the current compilation issue? It's availalbe [here](https://logs.ofborg.org/?key=nixos/nixpkgs.243298&attempt_id=dd7df179-46f8-4856-b9a7-dcd33c46fa4c).

reckenrode commented:
> Thanks for the link! Do you have any idea what can be done with the current compilation issue? It's availalbe [here](https://logs.ofborg.org/?key=nixos/nixpkgs.243298&attempt_id=dd7df179-46f8-4856-b9a7-dcd33c46fa4c).

portmidi doesn’t propagate its build inputs, so you don’t have to rebuild it. It also appears I was wrong about the stdenv. I thought Qt would provide an 11.0 SDK one, but it’s apparently still the default 10.12 SDK stdenv. Using the 11.0 SDK stdenv directly allows it to build for me. I apologize for not doing more testing (aside from confirming the expression for portaudio). Here’s a diff with the changes I made to get it to build for x86_64-darwin.

```diff
diff --git a/pkgs/applications/audio/musescore/default.nix b/pkgs/applications/audio/musescore/default.nix
index 07a9e36a0ab..da0c8110a66 100644
--- a/pkgs/applications/audio/musescore/default.nix
+++ b/pkgs/applications/audio/musescore/default.nix
@@ -31,10 +31,11 @@
 }:
 
 let
+  stdenv' = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
   # portaudio and portmidi propagate Darwin frameworks. Rebuild them using the
   # 11.0 stdenv from Qt and the 11.0 SDK frameworks.
   portaudio' = if stdenv.isDarwin then portaudio.override {
-      inherit stdenv;
+      stdenv = stdenv';
       inherit (darwin.apple_sdk_11_0.frameworks)
         AudioUnit
         AudioToolbox
@@ -43,17 +44,7 @@ let
         Carbon
       ;
     } else portaudio;
-  portmidi' = if stdenv.isDarwin then portmidi.override {
-      inherit stdenv;
-      inherit (darwin.apple_sdk_11_0.frameworks)
-        Carbon
-        CoreAudio
-        CoreFoundation
-        CoreMIDI
-        CoreServices
-      ;
-    } else portmidi;
-in stdenv.mkDerivation rec {
+in stdenv'.mkDerivation rec {
   pname = "musescore";
   version = "4.1.0";
 
@@ -101,7 +92,7 @@ in stdenv.mkDerivation rec {
     libsndfile
     libvorbis
     portaudio'
-    portmidi'
+    portmidi
     flac
     qtbase
     qtdeclarative
```

Unfortunately, it fails during install phase. At least it’s not SDK related ….

```console
CMake Error at src/app/cmake_install.cmake:87 (file):
  file INSTALL cannot find
  "/private/tmp/nix-build-qtbase-5.15.9.drv-0/qtbase-a196623/$(out)/$(qtQmlPrefix)":
  No such file or directory.
Call Stack (most recent call first):
  src/cmake_install.cmake:42 (include)
  cmake_install.cmake:57 (include)


FAILED: CMakeFiles/install.util
cd /tmp/nix-build-musescore-4.1.0.drv-0/source/build && /nix/store/q6cc8rhv9v8ji23xw3d8rp2ns0cclzc1-cmake-3.26.4/bin/cmake -P cmake_install.cmake
ninja: build stopped: subcommand failed.
```

doronbehar commented:
@ofborg build musescore

Thanks again for making the effort @reckenrode :). I think I've found the issue with upstream's cmake files, and that this is fixable with a patch. I added a patch, and I'll continue from here on with Ofborg I guess :), unless you are willing to keep helping.

reckenrode commented:
I can take a look later if you’re still having issues. I’m also interested in the CMake changes because it does seem like qmake in nixpkgs has retained paths to its build directory.

drupol commented on pkgs/applications/audio/musescore/default.nix:
```diff
@@ -54,6 +54,9 @@ in stdenv'.mkDerivation rec {
     sha256 = "sha256-CqW1f0VsF2lW79L3FY2ev+6FoHLbYOJ9LWHeBlWegeU=";
   };
+  patches = [
+    ./darwin-fix-install.patch
```
Can you please add a comment on why this patch is required?

doronbehar replied:
> Can you please add a comment on why this patch is required?

I will, once I see that it works.

reckenrode commented:
I was able to get it to build, but I ran into a few issues.

* It hardcodes x86_64 by default. I had to apply a fix.
* The app bundle should be installed to `$out/Applications`.
* The wrapper args need tweaked to distinguish between Darwin and Linux.
* It was failing to open due to missing `QtQuick.Dialogs`. I had to add `qtquickcontrols` as a `buildInput`.
* It appears there is some third-party dep stuff installed to `$out/lib` and `$out/include`. Is that needed?

The following diff resolves these issues (except the last one). I didn’t test except to confirm that it launched successfully.

```diff
diff --git a/pkgs/applications/audio/musescore/default.nix b/pkgs/applications/audio/musescore/default.nix
index 56185fccf0d..4b344516c11 100644
--- a/pkgs/applications/audio/musescore/default.nix
+++ b/pkgs/applications/audio/musescore/default.nix
@@ -20,6 +20,7 @@
 , qtdeclarative
 , qtgraphicaleffects
 , flac
+, qtquickcontrols
 , qtquickcontrols2
 , qtscript
 , qtsvg
@@ -70,7 +71,8 @@ in stdenv'.mkDerivation rec {
 
   qtWrapperArgs = [
     # MuseScore JACK backend loads libjack at runtime.
-    "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ libjack2 ]}"
+    "--prefix ${lib.optionalString stdenv.isDarwin "DY"}LD_LIBRARY_PATH : ${lib.makeLibraryPath [ libjack2 ]}"
+  ] ++ lib.optionals (!stdenv.isDarwin) [
     # There are some issues with using the wayland backend, see:
     # https://musescore.org/en/node/321936
     "--set-default QT_QPA_PLATFORM xcb"
@@ -100,6 +102,7 @@ in stdenv'.mkDerivation rec {
     qtbase
     qtdeclarative
     qtgraphicaleffects
+    qtquickcontrols
     qtquickcontrols2
     qtscript
     qtsvg
@@ -110,6 +113,16 @@ in stdenv'.mkDerivation rec {
     alsa-lib
   ];
 
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace build/cmake/SetupBuildEnvironment.cmake \
+      --replace 'set(CMAKE_OSX_ARCHITECTURES x86_64)' 'set(CMAKE_OSX_ARCHITECTURES ${stdenv.hostPlatform.darwinArch})'
+  '';
+
+  postInstall = lib.optionalString stdenv.isDarwin ''
+    mkdir -p "$out/Applications"
+    mv "$out/mscore.app" "$out/Applications/mscore.app"
+  '';
+
   passthru.tests = nixosTests.musescore;
 
   meta = with lib; {
```

reckenrode commented:
I did a search for a file I could open. Playback works too.

doronbehar commented:
Thanks again for the help @reckenrode . Regarding the `CMAKE_OSX_ARCHITECTURES x86_64` thing: So you tested this build on an m1 aarch64 darwin machine? I wonder whether we can / should contribute this fix, and my patch upstream.

Also, regarding the `postInstall`, Do you get eventually also a `$out/bin/mscore` binary? Is `$out/Applications/mscore.App` needed in addition to `$out/bin/mscore`?

doronbehar commented:
Oh and BTW I was wondering what kind of error do you get when `sandboxProfile` is not set to that `qtbase.__propagatedSandboxProfile`...

doronbehar commented on pkgs/applications/audio/musescore/default.nix:
```diff
@@ -1,65 +1,140 @@
+  # Upstream assumes the user has x86_64, and not aarch64
+  postPatch = lib.optionalString stdenv.isDarwin ''
+    substituteInPlace build/cmake/SetupBuildEnvironment.cmake \
+      --replace 'set(CMAKE_OSX_ARCHITECTURES x86_64)' 'set(CMAKE_OSX_ARCHITECTURES ${stdenv.hostPlatform.darwinArch})'
```
I wonder whether this can be set with a `cmakeFlag` (question for @reckenrode ).

reckenrode replied:
There is `BUILD_MACOS_APPLE_SILICON`, which can be set to `ON`. It works by clearing out `CMAKE_OSX_ARCHITECTURES`, which (as I understand it) means to build for the native host only. That should also work normally, but I’m not sure what it would do in a cross build.

I tried cross-building musescore (because x86_64-darwin to aarch64-darwin should work), but it appears waf needs an emulator for an aarch64-darwin cross-build, and there isn’t one available.

doronbehar replied:
> There is `BUILD_MACOS_APPLE_SILICON`, which can be set to `ON`. It works by clearing out `CMAKE_OSX_ARCHITECTURES`, which (as I understand it) means to build for the native host only. That should also work normally, but I’m not sure what it would do in a cross build.

Cross compilation is a very high ideal for this package IMO. I removed that `substituteInPlace` in favor of the mentioned cmake flag which I think will survive updates better, especially if upstream will rewrite/rename their cmake files.

reckenrode commented:
> Thanks again for the help @reckenrode . Regarding the `CMAKE_OSX_ARCHITECTURES x86_64` thing: So you tested this build on an m1 aarch64 darwin machine?

I tested on an M1 Max MBP.

> I wonder whether we can / should contribute this fix, and my patch upstream.

I assume such a patch would need to make copying the QML files conditional (e.g., on `MUE_USE_SYSTEM_QML`) because upstream probably wants to continue to embed them along with the Qt frameworks.

> Also, regarding the `postInstall`, Do you get eventually also a `$out/bin/mscore` binary? Is `$out/Applications/mscore.App` needed in addition to `$out/bin/mscore`?

There’s no `$out/bin/mscore`. It only creates the app bundle. If you also want `$out/bin/mscore`, you’ll need to symlink `$out/applications/mscore.app/Contents/MacOS/mscore` to `$out/bin/mscore`.

I also wanted to note there are the following things in `$out`. These don’t look related or required by musescore. If not, can they removed from `$out`?

```
result/include:
gmock/  gtest/  kddockwidgets/  opus/

result/lib:
cmake/  libgmock.a  libgmock_main.a  libgtest.a  libgtest_main.a  libkddockwidgets.a  libopus.a  pkgconfig/
```

reckenrode commented:
> Oh and BTW I was wondering what kind of error do you get when `sandboxProfile` is not set to that `qtbase.__propagatedSandboxProfile`...

One of the Qt applications using during the build process will crash during when the sandbox is enabled. See #234122 for details.

In theory, the profile should be propagated to anything using qtbase. It doesn’t seem to work though (see #237458).

doronbehar committed:
musescore: remove unneeded bundled libraries and headers from $out

doronbehar commented:
> I also wanted to note there are the following things in `$out`. These don’t look related or required by musescore. If not, can they removed from `$out`?

Defninitly, these are not darwin specific. I removed them.

> There’s no `$out/bin/mscore`. It only creates the app bundle. If you also want `$out/bin/mscore`, you’ll need to symlink `$out/applications/mscore.app/Contents/MacOS/mscore` to `$out/bin/mscore`.

OK good, I created that symlink.

> I assume such a patch would need to make copying the QML files conditional (e.g., on `MUE_USE_SYSTEM_QML`) because upstream probably wants to continue to embed them along with the Qt frameworks.

Yes I agree. I opened an issue there: https://github.com/musescore/MuseScore/issues/18665 .

Other then that, I squashed all the darwin related changes to 1 commit, and I'm pretty sure it should work. Thanks once more @reckenrode for all the help.

SuperSandro2000 commented on pkgs/applications/audio/musescore/default.nix:
```diff
@@ -1,65 +1,146 @@
 
-mkDerivation rec {
+let
+  stdenv' = if stdenv.isDarwin then darwin.apple_sdk_11_0.stdenv else stdenv;
```
```suggestion
  stdenv' = darwin.apple_sdk_11_0.stdenv;
```
the implementation is already doing this

doronbehar replied:
> the implementation is already doing this

I thought so too but I wasn't sure. Thanks!

doronbehar replied:
Wait no, I get "unsupported platform" errors with this change, because `darwin.apple_sdk_11_0.stdenv.isDarwin` is always `true`.

toonn replied:
Do you need the `stdenv'` at all if the package is `apple_sdk_11_0.callPackage`?

doronbehar replied:
> Do you need the `stdenv'` at all if the package is `apple_sdk_11_0.callPackage`?

The `portaudio` package is not defined using `darwin.apple_sdk_11_0.callPackage` if that's what you meant. How these things should be composed is discussed in general in https://github.com/NixOS/nixpkgs/issues/242666 .

reckenrode replied:
> Do you need the `stdenv'` at all if the package is `apple_sdk_11_0.callPackage`?

portaudio uses Qt’s callPackage, which doens’t appear to provide the 11.0 SDK stdenv. The Qt `callPackage` is needed to make sure the correct Qt dependencies are provided to the package.

> the implementation is already doing this

Only if you use `darwin.apple_sdk_11_0.callPackage`. The 11.0 SDK stdenv itself is not defined conditionally (except to pass through the existing 11.0 SDK stdenv on aarch64-darwin).

drupol commented:
There's one missing test and it's ready to merge after that. Nice team work in this PR !

drupol commented:
Ok enough waiting.

yurrriq commented:
> cc @yurrriq per https://github.com/NixOS/nixpkgs/pull/23204 .

@doronbehar et al. thanks for this PR! Six years ago :open_mouth: I just did a naive port of the Homebrew... recipe? to get a quick and dirty (and binary) expression for darwin. I'm glad to see this finally getting fixed.
